adds c.introduceTopLevel

xeno-by opened the pull request:
The first in the family of mutators for the global symbol table,
`introduceTopLevel` is capable of creating synthetic top-level
classes and modules.

The addition of nme.EMPTY_PACKAGE_NAME is necessary to let
programmers insert definitions into the empty package. That's explicitly
discouraged in the docs, but at times might come in handy.

This patch introduce workarounds to avoid incompatibilities with SBT.
First of all SBT doesn't like VirtualFiles having JFile set to null.
Secondly SBT gets confused when someone depends on synthetic files
added by c.introduceTopLevel.

Strictly speaking these problems require changes to SBT, and that will be
done later. However the main target of the patch is paradise/macros,
which needs to be useful immediately, therefore we apply workarounds.

xeno-by committed:
adds c.introduceTopLevel

The first in the family of mutators for the global symbol table,
`introduceTopLevel` is capable of creating synthetic top-level
classes and modules.

The addition of nme.EMPTY_PACKAGE_NAME is necessary to let
programmers insert definitions into the empty package. That's explicitly
discouraged in the docs, but at times might come in handy.

This patch introduce workarounds to avoid incompatibilities with SBT.
First of all SBT doesn't like VirtualFiles having JFile set to null.
Secondly SBT gets confused when someone depends on synthetic files
added by c.introduceTopLevel.

Strictly speaking these problems require changes to SBT, and that will be
done later. However the main target of the patch is paradise/macros,
which needs to be useful immediately, therefore we apply workarounds.

xeno-by commented:
review @paulp @adriaanm

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1244/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1244/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1962/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1962/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1245/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1245/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1964/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1964/

paulp commented on src/compiler/scala/reflect/macros/runtime/Synthetics.scala:
```diff
@@ -0,0 +1,66 @@
+        if (name.isTermName) mirror.staticModule(name.toString)
+        else mirror.staticClass(name.toString)
+      } catch {
+        case MissingRequirementError(_) => NoSymbol
```
Must we duplicate this wrapper? How about making 'def wrapMissing' final instead of private.

xeno-by replied:
Done

paulp committed:
Made Symbol#associatedFile always return non-null

So we don't have to clutter everything with null checks.

paulp commented:
I haven't tested this except to see that it compiles, but I feel this code was overcomplicated so I made some tweaks which I offer here f543b45ac842  .

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1263/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1263/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1982/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1982/

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
also review @paulbutcher

xeno-by commented:
Incorporated Paul's and Paul's feeback. New code should be less complicated and more pleasant to use with type macros.

In the simplest case, you just do `c.topLevelRef(name) orElse c.introduceTopLevel(packageSym, generate())`, and you no longer need to generate reference to packages or to synthesized classes.

If you need to re-generate when something changes during the interactive session (in the IDE, in Ensime), then you attach that something to the result in `generate` and then:

```
c.topLevelDef(name).attachments.get[YourAttachment] match {
  case Some(payload) if payload.unchanged => c.topLevelRef(name)
  case _ => c.introduceTopLevel(packageSym, generate())
}
```

I also removed `c.existsOnClasspath`, since this doesn't seem to be a very useful method and can be easily emulated by a combo of `c.staticClass`/`c.staticModule` and `c.topLevelDef.isEmpty`.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1990/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1990/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1278/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1278/

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
@paulp It looks like your changes crash Scaladoc. I'll try to figure out what's going on.

In the meanwhile I rolled back the `associatedFile` part, because I don't think that replacing `associatedFile ne null` checks with `associatedFile.exists` checks will work. That's because of this very patch, which adds an API to introduce valid compilation units with non-existent underlying files.

paulbutcher commented:
I'm not competent to comment on any of the implementation details, so I'll restrict myself to commenting on the API. I like the basic structure and the cleaner way of handling package definitions. I'm not sure that I understand how I can use attachments to deal with the need to regenerate though. In my case, for example, I need to regenerate a mock definition if the original class upon which it's based has changed. What will I attach to my mock to allow me to achieve this?

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1991/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1991/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1279/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1279/

paulp commented:
@xeno-by There are already valid compilation units without underlying files. What/why is associating "validity" with "has underlying file" ?

paulp commented:
I want to commit those associatedFile changes regardless of anything else. The check ".exists" should be "!= NoAbstractFile" if .exists was a problem. I actually meant ".exists" in the NoSymbol sense, but I didn't realize there was divergence underneath.

xeno-by commented:
@paulp Ah now I see. I will apply your changes.

xeno-by commented:
@paulbutcher It's a hard question. I'd probably look up a def tree for the type being mocked and notice its `id` (cannot be obtained through the reflection API at the moment, needs a cast). Then I'd write a class `case class OriginalAttachment(id: Int)` that would carry that id and do `mockClassDef.attachments update OriginalAttachment(id)`.

@paulp How does one detect that a certain definition has changed during the interactive scalac session?

paulp commented:
@xeno-by There's some complicated machinery which I don't know if it even works, but see "def invalidateClassPathEntries", "def reSync", etc. in Global. If sbt doesn't use it then it doesn't work. It was written for use by sbt.

xeno-by commented:
@paulp Now after I also changed `Symbol.sourceFile`, `docs.lib` works fine. I don't see how we can change `sourceFile` to be non-null though. Who knows who can rely on its current behavior.

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1280/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1280/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1992/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1992/

xeno-by commented:
rebased against `master`. PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1286/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1286/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2000/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2000/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2017/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2017/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1303/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1303/

paulp commented:
This looks acceptable to me.
